### ISSUE TITLE
Unify Citus executors

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#1273]
Proposal for a new executor to replace all other executors, but with the initial goal of replacing the real-time and multi-shard router executors with one that addresses most of their shortcomings.

## Why a new executor?

As we're expanding our SQL support for multi-shard queries (#1204), we expect Citus to power more applications that query across shards. However, in current Citus workloads the multi-shard executors have several problems:

- Not transactional - #1119
- Excessive number of connections, TIME_WAIT issues (real-time, multi-shard router)
- Might not be able to complete due to number of shards (multi-shard router) - #709
- No connection reuse (real-time, multi-shard router)
- Does not consider query time / data imbalances (real-time, multi-shard router)
- Requires complex pgbouncer configuration (real-time, multi-shard router)
- No query parameter support - #1179
- Deployment issues (task-tracker) - #564, requires .pgpass on workers
- Poor performance (task-tracker)
- Not well-understood (task-tracker)
- Implementation issues (mainly task-tracker) - #716, #255, #353, #69
- Need to manually select executor - #444
- Technical debt - #292, #294, #296, #1121, #1260

We currently have 4 different query executors: router, real-time, task-tracker, and multi-shard router (DDL/INSERT..SELECT/master_modify_multiple_shards), all with different properties. While this could be considered a way of addressing different trade-offs for different types of queries, the main trade-off for most of them was expediency. Actual technical differences could be represented through a small number of parameters in a more general executor.

## General Approach 

The unified executor establishes a small connection pool to each worker at the start of a query and assigns tasks to connections as they become available. Each connection has its own queue for tasks that need to be assigned to it because of preceding writes in the same transaction, and other tasks are added to a general queue. Once a connection is established and optionally completes BEGIN, it can either take a task from its own queue, or one from the general queue, and keeps doing so until all tasks are done.

## Advantages

By using small, possibly cached connection pools, we can reduce or even eliminate connection churn from multi-shard SELECT queries. We can also support SELECT and DDL with a much larger number of concurrent queries and much larger shard count, which also enables new sharding approaches (e.g. range-distribution, multi-level sharding). We can also allow users to adjust the pool size to increase or decrease resources used by a particular query.

Another advantage of this model is that if a connection takes a while to become available, because one task is much bigger than others, because of a network timeout, or because it takes a while for an intermediate pgbouncer to assign a new connection, then the remaining tasks can still finish on the available connections, which can speed up queries and avoids cases in which pgbouncer can get stuck due to concurrent queries requiring >1 connections to complete (except during COPY, which always requires >1 connections). 

This model also allows us to easily batch multiple commands that need to go to the same worker together by taking multiple items from the queue before starting a new task. This could significantly speed up many multi-shard DDL/UPDATE/DELETE commands by avoiding round-trips or opening many connections, which allows us to release locks earlier. Moreover, the ability to batch a large number of tasks also puts us in a good position to eliminate task-tracker. With an event-driven executor that can do batching and uses a fixed a connection pool, the motivations for having a separate task scheduler on the workers become largely obsolete.

Overall, the new executor can reduce execution times for multi-shard queries, allow transactional multi-shard SELECTs, alleviate the connection churn that causes active customer pain, enable high shard counts and by extension new sharding approaches, and simplify our code.

## Drawbacks

Compared to the current real-time executor, the new executor will have a few additional knobs. The pool size used for various commands needs to be tuned. We want to give a good initial impression by using all available cores, but we probably want to use fewer connections in production to allow for greater concurrency. It's possible that we want separate GUCs for administrative commands and SELECT queries. We might also want a GUC for the number of connections that are cached throughout the session.

Batch sizes are also tricky because we would prefer not to put commands with high execution times in a batch, but we don't know in advance which commands have high execution times. We may want to control batching based on the type of task (e.g. always batch worker_fetch_partition_file, never batch a SELECT task) and the number of tasks in the queue.

## Implementation

Notes on implementing the new executor.

### Algorithm 

The algorithm for the executor will look roughly like:

```

/*
 * Get all previously established connections to workers. 
 * 
 * Add tasks that read from placements that have already been modified to the
 * taskQueue of the modifying connection.
 *
 * Open additional connections if there are fewer than defaultPoolSize connections
 * in a pool, unless all tasks can are already mapped to a connection.
 */
connections = CreateConnectionPoolsForTasks(defaultPoolSize, allTasks, &unassignedTasks);
pendingTaskCount = CountTasks(allTasks);

while (pendingTaskCount > 0) {

  WaitForEvents(connections);

  foreach connection in connections
    switch(connection->state)
      case CONNECTION_ESTABLISHING:
        ConnectionEstablishmentStateMachine(connection);
        break;

      case TASK_DONE:
        TaskCompleted(connection);
        pendingTaskCount--;

      case READY_FOR_NEXT_TASK:
        /* could also get a batch of tasks here*/

        task = PopTaskFromQueue(connection->taskQueue);

        if (task == NULL)
          task = PopTaskForWorker(unassignedTasks, connection);

        if (task == NULL)
          ReleaseConnection(connection);

        StartTaskOnConnection(task, connection);

      case TASK_RUNNING:
        TaskStateMachine(connection->currentTask);

}
```

### Data structures

To address task-tracker (MapMerge) queries, we need to consider dependencies between tasks. A general representation of a distributed query plan is a directed acyclic dependency graph in which a task can start once all of its dependencies are done. A possible implementation which ensures that we execute tasks in the right order could be to keep track of an `unresolvedDependenciesCount` in each task. `PopTaskFromQueue` and `PopTaskForWorker` skip over a task as long as `unresolvedDependenciesCount` is greater than 0. `TaskCompleted` lowers `unresolvedDependenciesCount` by 1 for each dependent task.

We currently have the notion of a task with several task placements. To execute a SELECT query, we try the placements in order until one succeeds, opening a new connection for each placement. In the new executor, we want to be able to map a task to a specific connection, while different placements map to different connections. A possible implementation would be to add the same task struct to all applicable queues, but keep a pointer to the current placement (similar to the current `TaskExecution` struct). `PopTaskFromQueue` and `PopTaskForWorker` skip over a task as long as the current placement belongs to a different worker or if the task completed once.  `TaskCompleted` (or `TaskFailed`) increments the current placement pointer to cycle through all placements.

We should also be able to handle DML and INSERT..SELECT commands through the same executor. An additional constraint in DML commands and INSERT commands in particular is that they need to be executed in the same order across the placements, since they might obtain conflicting row-level locks. A possible implementation would be to follow the same approach as for SELECT, but complete (or fail) the task for every placement. UPDATE/DELETE/DDL commands across multiple placements could be executed out-of-order, since we obtain exclusive shard locks before executing them. In that case, the current placement is irrelevant and `PopTaskFromQueue` and `PopTaskForWorker` can always return the task if it has not been executed for the placement. It might make sense to have a flag indicating the execution type in each task struct.

### Fetching results

A key implementation difference between the current executors lies in the way they fetch results. Task-tracker transmits result files to the master over a single connection, real-time executor runs `COPY (SELECT ..) TO STDOUT` commands across many connections and stores the output in files, multi-shard router executor runs queries directly and collects results in a tuple store.

It might make sense to directly collect results in a tuple store in all cases, though we can also support multiple modes.

### Transactional MapMerge queries

Ideally, the tasks in a MapMerge query should be executed in transaction blocks, rather than performing intermediate commits. To do so, we do need to make sure that when an intermediate table ("fragment") is created, subsequent queries that read from that fragment are assigned to the connection that created it. Moreover, fragments should be considered "co-located" with the placements with which they are joined. This is not very complicated to add to the connection API, but a simpler implementation would be to only allow MapMerge queries outside of a transaction block separately commit all tasks.

Task-tracker stores intermediate files in a job directory and intermediate tables in a job schema and removes both at the end of the statement by calling `task_tracker_cleanup_job`. If we implement transaction support for MapMerge queries, then it might make sense to use temporary tables instead of a job schema to simplify the cleanup logic and set a resource owner callback to remove the job directory.

**Comment by anarazel** [comment]
@marcocitus I think this makes a lot of sense.

One additional design goal I'd like to propose is that the executor should be able to run to "partial completion". I.e. return to the caller as soon as some rows are ready to be returned.   The current behaviour requires to temporarily store all returned data on the issuing node. That's bad because it means it'll take forever to return data of big queries to the user, even if no group bys etc are involved, and because it can require tremendous intermediary storage on the master.  If we e.g. started returning rows to the calling custom scan whenever they're available, re-aggregation queries can avoid storing all data, and non re-aggregating queries (e.g. COPY OUT) can immediately return rows.

I think there's probably a need for a "transparent restart" GUC, because a lot of optimizations are possible when that's not required, and a good chunk of the citus userbase doesn't need it.


One question: Once we get to the point that we can have multiple pull-to-master style queries in one bigger query, do you think it's ok to let those connection limits / pools to apply to only individual queries in that?

**Comment by marcocitus** [comment]
> Once we get to the point that we can have multiple pull-to-master style queries in one bigger query, do you think it's ok to let those connection limits / pools to apply to only individual queries in that?

I think things get pretty weird if different CTEs/subqueries work on different snapshots. We may have to set the isolation level to repeatable read and use the same connections for the same co-located placements.

**Comment by anarazel** [comment]
> > Once we get to the point that we can have multiple pull-to-master style queries in one bigger query, do you think it's ok to let those connection limits / pools to apply to only individual queries in that?
> 
> I think things get pretty weird if different CTEs/subqueries work on different snapshots. We may have to set the isolation level to repeatable read and use the same connections for the same co-located placements.

That works easily enough for router, but for real-time/tt type queries
that won't work well. Nor are they currently consistent...

For router style queries it's usually not very interesting to pipeline
query processing that much anyway, imo, so I think my concern here
really is mostly about non router queries.

**Comment by marcocitus** [comment]
> That works easily enough for router, but for real-time/tt type queries that won't work well. Nor are they currently consistent...

Real-time queries are locally consistent at least. If I do a select with several subqueries, either all subqueries see a preceding insert or none do. I'd be hesitant to move away from these semantics. Imagine how awkward joins get.

**Comment by jasonmp85** [comment]
> But they were all of them deceived, for another executor was made: in the land of Amsterdam, in the waters of Herengracht, the dark lord @marcocitus forged, in secret, a master executor to control all others. And into this executor he poured his cruelty, his malice,and his will to dominate all queries.

👍

**Comment by marcocitus** [comment]
I'd consider this done. We should consider dropping the task-tracker in favour of scheduling tasks via the adaptive executor.
